draft-ietf-gnap-core-protocol.md

```diff
@@ -2368,108 +2370,199 @@ unknown_request
-user code cannot be associated with a currently active request, the AS
-MUST display an error to the RO and MUST NOT attempt to redirect the
-RO back to any client instance even if a [redirect is supplied](#request-interact-callback-redirect).
+to redirect the RO back to any client instance even if a [redirect finish method is supplied](#request-interact-callback-redirect).
+If the URI is not hosted by the AS directly, the means of communication between
+the AS and this URI are out of scope for this specification.
+
+The client instance MUST NOT modify the URI when launching it,
+in particular the client instance MUST NOT add any parameters to the URI.
+The URI MUST be reachable from the end user's browser, though
+the URI MAY be opened on a separate device from the client instance
+itself. The URI MUST be accessible from an HTTP GET
+request and MUST be protected by HTTPS or equivalent means.
+
+### Interaction at the User Code URI {#interaction-usercode}
+
+When the end user is directed to enter a short code through the ["user_code"](#response-interact-usercode)
+mode, the client instance communicates the user code to the end-user and 
+directs the end user to enter that code at an associated URI.
+This mode is used when the client instance is not able to facilitate launching
+an arbitrary URI. The associated URI could be statically configured with the client instance or 
+communicated in the response from the AS, but the client instance
+communicates that URL to the end user. As a consequence, these URIs SHOULD be short.
+
+In many cases, the URI indicates a web page hosted at the AS, allowing the
+AS to authenticate the end user as the RO and interactively provide consent.
+If the URI is hosted by the AS,
+the AS MUST determine the grant request being referenced from the user code.
+If the user code cannot be associated with a currently active
+request, the AS MUST display an error to the RO and MUST NOT attempt
+to redirect the RO back to any client instance even if a [redirect finish method is supplied](#request-interact-callback-redirect).
+If the interaction component at the user code URI is not hosted by the AS directly, the means of communication between
+the AS and this URI, including communication of the user code itself, are out of scope for this specification.
 
-The user code URL MUST be reachable from the RO's browser, though
-note that the RO MAY open the URL on a separate device from the client instance
-itself. The user code URL MUST be accessible from an HTTP GET request,
-and MUST be protected by HTTPS or equivalent means.
+When the RO enters this code at the user code URI,
+the AS MUST uniquely identify the pending request that the code was associated
+with. If the AS does not recognize the entered code, the interaction component MUST
+display an error to the user. If the AS detects too many unrecognized codes
+enter attempts, the interaction component SHOULD display an error to the user.
```

Comment:
the AS may also decide to take any appropriate action in case there are too many errors (like freeze temporarily). If there's a maximum number of attempts, it probably should be displayed to the user before that kind of decision is taken